api/v1beta1/kustomization_types.go

```diff
@@ -254,13 +256,35 @@ func SetKustomizationReadiness(k *Kustomization, status metav1.ConditionStatus,
 
 // KustomizationNotReady registers a failed apply attempt of the given Kustomization.
 func KustomizationNotReady(k Kustomization, revision, reason, message string) Kustomization {
+	RemoveKustomizationReconciling(&k)
+	SetKustomizationStalled(&k, metav1.ConditionTrue, reason, message)
 	SetKustomizationReadiness(&k, metav1.ConditionFalse, reason, trimString(message, MaxConditionMessageLength), revision)
 	if revision != "" {
 		k.Status.LastAttemptedRevision = revision
 	}
 	return k
 }
 
+// SetKustomizeStalled sets the StalledCondition, ObservedGeneration on the Kustomization.
+func SetKustomizationStalled(k *Kustomization, status metav1.ConditionStatus, reason, message string) {
+	meta.SetResourceCondition(k, meta.StalledCondition, status, reason, trimString(message, MaxConditionMessageLength))
+	k.Status.ObservedGeneration = k.Generation
```

Comment:
Should this be set on `Stalled`, or on reaching `Ready`? Or do we want to record it in the `ObservedGeneration` field of the `metav1.Condition` Status, but only move the one in the `Status` on a successful run?

Reply:
`kstatus` seems to return an `InProgress` status once `status.ObserveredGeneration` != `metadata.generation` regardless of the conditions present. So I thought we would want to set just incase,

Reply:
Good catch! Which brings us to a new question: does every error mean that the reported condition status is `Stalled`, and `kstatus` should thus stop waiting because `ObservedGeneration == Generation` or `Status == Stalled`?

If we for example run into a transient error (e.g. transient resource lookup failure that returned an error from the K8s API server). This would mean that `kstatus` would stop waiting for the resource to reach the expected state if we would set both `ObservedGeneration` and `Stalled`, while if we would set `Ready==False` or left the `ObservedGeneration` untouched, it would continue to wait for the next reconciliation to (hopefully) succeed: https://github.com/kubernetes-sigs/cli-utils/blob/e351b2bc43cec2107ba1d874c3dec54fd0956c59/pkg/kstatus/status/generic.go#L39-L54

Reply:
> If we for example run into a transient error (e.g. transient resource lookup failure that returned an error from the K8s API server).

Hmm. We probably have to identify transient errors and real errors, then decide how best to handlethe transient. I think we should set it `Reconciling` status for transient error and `Ready=false`.
 kstatus takes `ObservationGeneration != generated` to mean that the controller might not be running or unable to update the resource. 
 
 Plus a generic message is when `ObservationGeneration != generated`. If we use `Reconciling` condition, kstatus sets the reason and message from the message.